### PR TITLE
scala API

### PR DESCRIPTION
**Pull request by joelgrus**
at a high level

* moved the swig bindings into `edu.cmu.dynet.internal` (didn't otherwise touch them)
* created a scala-y API that doesn't expose any of the C++ internals
* mostly wiped out the `DyNetScalaHelpers` (most of them are unnecessary under the new API) and renamed the few left to `Utilities`
* moved the project from dynet/swig to dynet/contrib/swig
* updated the documentation

I think at this point the scala API is basically feature complete (relative to the C++ one).

still haven't touched the "more functional" piece. I have been playing around with it in a half-baked way, we can discuss.

### CODE REVIEW

File: contrib/swig/README.md [hunk 1]
```diff
@@ -170,108 +179,50 @@ iter = 29, loss = 8.881784E-16
...
-`IntVector`, `FloatVector`, `ExpressionVector`, and so on. Each has a 
-no-argument constructor, a `capacity: Int` constructor, and a
-`elems: java.util.Collection[T]` constructor (for the relevant type `T`).
-
-`DyNetScalaHelpers` contains implicit conversions to `Collection[T]` 
-for the corresponding `Seq` types, so that you can do things like
+DyNet does a lot behind the scenes with C++ `std::vector<>`s. 
+In Scala there are `IntVector`, `FloatVector`, `UnsignedVector`, and
+`ExpressionVector` classes that thinly wrap these C++ vectors. 
+They all implemented `IndexedSeq` so that they're pretty easy to work
```
Comment (jayantk) [inline contrib/swig/README.md]: typo: implement

File: contrib/swig/README.md [hunk 2]
```diff
@@ -170,108 +179,50 @@ iter = 29, loss = 8.881784E-16
 
 ## Usage
 
-The Scala version of DyNet is intended to work mostly like the
-C++. However, there are a few things to watch out for, which are
-documented below.
-
-### Imports
+The current Scala API works mostly like the C++ API, with the following
+differences.
```
Comment (jayantk) [inline contrib/swig/README.md]: Perhaps mention that functions have been renamed from underscores to camelcase?

File: contrib/swig/src/main/scala/edu/cmu/dynet/Dim.scala [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  def set(i: Long, s: Long): Unit = dim.set(i, s)
+  def get(i: Long): Long = dim.get(i)
+  def size(i: Long): Long = dim.size(i)
+
+  def deleteDim(i: Long): Unit = dim.delete_dim(i)
+
+  def transpose(): Dim = new Dim(dim.transpose())
+
+  /** We override `equals` so that `Dim` objects should be equal whenever all of their "dimension
+    * numbers" match.
```
Comment (jayantk) [inline contrib/swig/src/main/scala/edu/cmu/dynet/Dim.scala]: dimension sizes?

File: contrib/swig/src/main/scala/edu/cmu/dynet/Expression.scala [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+  private type UnaryTransform = internal.Expression => internal.Expression
+  private def unary(e: Expression, transformer: UnaryTransform) = {
+    e.ensureFresh()
+    new Expression(transformer(e.expr))
+  }
+
+  def exprMinus(e: Expression): Expression = unary(e, dn.exprMinus)
+  def exprPlus(e1: Expression, e2: Expression): Expression = binary(e1, e2, dn.exprPlus)
+  def exprPlus(e1: Expression, x: Float): Expression = unary(e1, e1 => dn.exprPlus(e1, x))
+  def exprPlus(x: Float, e2: Expression): Expression = exprPlus(e2, x)
```
Comment (jayantk) [inline contrib/swig/src/main/scala/edu/cmu/dynet/Expression.scala]: missing unary

File: contrib/swig/src/main/scala/edu/cmu/dynet/GRUBuilder.scala [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package edu.cmu.dynet
+
+/** Builder method for creating GRUs, as in the C++ code. For its public methods see
+  * [[edu.cmu.dynet.RNNBuilder]].
+  */
+class GRUBuilder private[dynet](private[dynet] val builder: internal.GRUBuilder)
+    extends RNNBuilder(builder) {
+
+  /** Create a new, empty LSTMBuilder. */
```
Comment (jayantk) [inline contrib/swig/src/main/scala/edu/cmu/dynet/GRUBuilder.scala]: GRUBuilder

File: contrib/swig/src/main/scala/edu/cmu/dynet/GRUBuilder.scala [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package edu.cmu.dynet
+
+/** Builder method for creating GRUs, as in the C++ code. For its public methods see
+  * [[edu.cmu.dynet.RNNBuilder]].
+  */
+class GRUBuilder private[dynet](private[dynet] val builder: internal.GRUBuilder)
```
Comment (jayantk) [inline contrib/swig/src/main/scala/edu/cmu/dynet/GRUBuilder.scala]: My preference for this name would be GruBuilder (at least, that's Google Java style). Same for LstmBuilder, etc. Let's ignore for now considering we're going to rethink RNNs though.

File: contrib/swig/src/main/scala/edu/cmu/dynet/Serialization.scala [hunk 7]
```diff
@@ -0,0 +1,56 @@
+package edu.cmu.dynet
+
+/** Class for persisting models and parameters to disk */
+class ModelSaver private[dynet](saver: internal.ModelSaver) {
+  def this(filename: String) { this(new internal.ModelSaver(filename)) }
+
+  def addModel(model: Model): Unit = saver.add_model(model.model)
+  def addParameter(p: Parameter): Unit = saver.add_parameter(p.parameter)
+  def addLookupParameter(p: LookupParameter): Unit = saver.add_lookup_parameter(p.lookupParameter)
+  def addLSTMBuilder(p: LSTMBuilder): Unit = saver.add_lstm_builder(p.builder)
```
Comment (jayantk) [inline contrib/swig/src/main/scala/edu/cmu/dynet/Serialization.scala]: Note that a capitalization rename would have to happen here too.

**Comment by jayantk** [comment]
LGTM! Looks great! It's a little annoying to have to write all the wrapper scala code, but code written using the new interface looks way more scala-y.

**Review by jayantk — APPROVED** [review]

File: contrib/swig/src/main/scala/edu/cmu/dynet/Expression.scala (hunk 4 above)
Comment (joelgrus) [reply contrib/swig/src/main/scala/edu/cmu/dynet/Expression.scala]: I was being cute and implementing it using the function above, but I changed it, because that's probably not a good idea

File: contrib/swig/src/main/scala/edu/cmu/dynet/GRUBuilder.scala (hunk 6 above)
Comment (joelgrus) [reply contrib/swig/src/main/scala/edu/cmu/dynet/GRUBuilder.scala]: IDEs make renaming fast, done

**Comment by joelgrus** [comment]
ok, addressed everything. I'll try to update PNP to account for these changes today
